### PR TITLE
Allow TypedDict key with literal type during construction

### LINKED ISSUE #7644: Literal type not accepted as key when constructing TypedDict
@b0g3r reported an issue with TypedDict keys not being accepted with literal types. Example code from https://github.com/python/mypy/issues/6262#issuecomment-538520768:

```py
from mypy_extensions import TypedDict
from typing_extensions import Literal, Final


class MyDict(TypedDict):
    field_name: int

d: MyDict
d = {'field_name': 1}  # OK

name_in_var = 'field_name'
d = {name_in_var: 1}  # Correct error

name_in_literal2: Literal['field_name'] = 'field_name'
d = {name_in_literal2: 1}  # Unexpected error

name_in_final: Final = 'field_name'
d = {name_in_final: 1}  # Unexpected error
```

### PR DESCRIPTION
**Pull request by JukkaL**
Also refactored an operation away from `mypy.plugins.common` so that it
is more cleanly reusable. Also kept an alias in the original location
to avoid breaking existing plugins.

Fixes #7644.

### CODE REVIEW

File: test-data/unit/check-typeddict.test [hunk 1]
```diff
@@ -1883,3 +1883,25 @@ assert isinstance(u2, Mapping)
...
+v = {num: ''}  # E: Incompatible types (expression has type "str", TypedDict item "num" has type "int")
+
+bad: Final = 2
+v = {bad: 3}  # E: Expected TypedDict key to be string literal
+union: Literal['num', 'foo']
+v = {union: 2} # E: Expected TypedDict key to be string literal
+num2: Literal['num']
+v = {num2: 2}
+bad2: Literal['bad']
+v = {bad: 2}  # E: Expected TypedDict key to be string literal
```
Comment (ilevkivskyi) [inline test-data/unit/check-typeddict.test]: I suppose this should be `bad2`, otherwise `bad2` would be unused in this test.

File: mypy/typeops.py [hunk 2]
```diff
@@ -417,3 +419,38 @@ def callable_type(fdef: FuncItem, fallback: Instance,
...
+    elif isinstance(typ, UnionType):
+        possible_literals = list(typ.items)
+    else:
+        possible_literals = [typ]
+
+    strings = []
+    for lit in get_proper_types(possible_literals):
+        if isinstance(lit, LiteralType) and lit.fallback.type.fullname() == 'builtins.str':
+            val = lit.value
+            assert isinstance(val, str)
```
Comment (b0g3r) [inline mypy/typeops.py]: It's not a review comment, just my curiosity: why is this needed?

Comment (JukkaL) [reply mypy/typeops.py]: Mypy thinks that the value could be an integer, for example, since it doesn't understand the `fullname()` check above.
